[c10d] use a promise to delay watchdog shutdown

c-p-i-o opened the pull request:
Summary:
We always need to give the heartbeat monitor thread time to write out flight recorder dumps. Otherwise, the watchdog thread kills the heartbeat monitor thread too fast before it has time to write out the Flight Recorder logs.
This change:
1. Removes the "sleep after exception" JK. We don't need to sleep for 8 minutes.
2. Use a promise between watchdog thread and heartbeat monitor thread to delay, at most, one minute to give Flight Recorder time to write out it's log on timeout.

Test Plan:
Tested on my local job and flight recorder successfully executed for the job.
https://fburl.com/mlhub/38fj5yne
The watchdog thread gives heartbeat thread time to write out the logs.

In the logs we see:
```
[trainer4]:I1023 17:39:29.755507 12592 ProcessGroupNCCL.cpp:1950] [PG ID 0 PG GUID 0(precheck) Rank 12] slept for 1647ms giving time for flight recorder dumps to finish.
```

Differential Revision: D64857928




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -1041,6 +1034,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   int waitTimeoutDumpInMilSec_;
 
+  // promise to coordinate flight recorder dump.
+  std::promise<bool> promiseFlightRecorderDump_;
```
nit: maybe use a void promise if we're not consuming the value?

fduwjj replied:
oh interesting, didn't know that.

d4l3k reviewed (approved):
LGTM -- would be nice to add a unit test if possible

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1950,14 +1951,21 @@ void ProcessGroupNCCL::watchdogHandler() {
+            auto status = promiseFlightRecorderDump_.get_future().wait_for(
+                std::chrono::milliseconds(waitTimeoutDumpInMilSec_));
+            if (status == std::future_status::timeout) {
+              LOG(INFO) << logPrefix() << "timed out after waiting for "
```
Do we want LOG(ERROR) or LOG(INFO)?

c-p-i-o replied:
I changed this to `ERROR`.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1950,14 +1951,21 @@ void ProcessGroupNCCL::watchdogHandler() {
+                        << computeDeltaMS(start, end) << "ms"
                         << " giving time for flight recorder dumps to finish.";
             }
+
```
nit: is this one expected?

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1385,8 +1384,10 @@ bool ProcessGroupNCCL::dumpDebuggingInfo() {
     LOG(INFO) << logPrefix() << "ProcessGroupNCCL dumping nccl trace to "
               << writer.getWriterTarget();
     writer.write(ncclTrace);
+    promiseFlightRecorderDump_.set_value(true);
     return true;
   }
+  promiseFlightRecorderDump_.set_value(false);
```
actually, is it better we just set_value inside monitoring thread?

c-p-i-o replied:
Good idea! Done!

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -64,13 +64,6 @@ static std::vector<std::string> TORCH_NCCL_ASYNC_ERROR_HANDLING = {
 static std::vector<std::string> TORCH_NCCL_DUMP_ON_TIMEOUT = {
     "TORCH_NCCL_DUMP_ON_TIMEOUT"};
 
-// TODO: remove this change after a safe rollout.
-// Control whether we sleep after an exception is thrown.
-// This change is temporary and is used to safely remove the current sleep that
-// exists after an exception is thrown.
-static std::vector<std::string> TORCH_NCCL_SLEEP_AFTER_EXCEPTION = {
-    "TORCH_NCCL_SLEEP_AFTER_EXCEPTION"};
```
also I would ensure this won't affect llama. Does llama job use JK knobs?

c-p-i-o replied:
No, llama doesn't use JKs. However, they do have DUMP_ON_TIMEOUT enabled. So, with this change, they will wait, at most one minute, on timed out jobs so that FR dumps can be written out.

fduwjj reviewed (approved):
LGTM

c-p-i-o committed:
[c10d] use a promise to delay watchdog shutdown (#138828)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/138828

We always need to give the heartbeat monitor thread time to write out flight recorder dumps. Otherwise, the watchdog thread kills the heartbeat monitor thread too fast before it has time to write out the Flight Recorder logs. This is resulting in Flight Recorder dumps missing for many jobs.

This change:
1. Removes the "sleep after exception" JK. We don't need to sleep for 8 minutes.
2. Use a promise between watchdog thread and heartbeat monitor thread to delay, at most, one minute to give Flight Recorder time to write out its log on timeout.

Test Plan:
Tested on my local job and flight recorder successfully executed for the job.
https://fburl.com/mlhub/38fj5yne
The watchdog thread gives heartbeat thread time to write out the logs.

In the logs we see:
```
[trainer4]:I1023 17:39:29.755507 12592 ProcessGroupNCCL.cpp:1950] [PG ID 0 PG GUID 0(precheck) Rank 12] slept for 1647ms giving time for flight recorder dumps to finish.
```

Reviewed By: d4l3k

Differential Revision: D64857928

wdvr commented:
@pytorchmergebot merge -f "stuck on macos job - force merging"
